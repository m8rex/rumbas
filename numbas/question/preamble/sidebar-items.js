window.SIDEBAR_ITEMS = {"struct":[["Preamble",""]]};