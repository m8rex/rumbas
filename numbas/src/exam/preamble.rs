use schemars::JsonSchema;
use serde::Deserialize;
use serde::Serialize;
use serde_with::skip_serializing_none;

#[skip_serializing_none]
#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
pub struct Preamble {
    pub js: String,
    pub css: String,
}
