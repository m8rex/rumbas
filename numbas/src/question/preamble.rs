use schemars::JsonSchema;
use serde::Deserialize;
use serde::Serialize;
use serde_with::skip_serializing_none;

#[skip_serializing_none]
#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
pub struct Preamble {
    pub js: String,
    pub css: String,
}

impl std::default::Default for Preamble {
    fn default() -> Self {
        Self { js: "".to_string(), css: "".to_string() }
    }
}