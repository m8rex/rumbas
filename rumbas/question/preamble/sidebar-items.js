window.SIDEBAR_ITEMS = {"enum":[["PreambleChange",""],["PreambleInputChange",""]],"struct":[["Preamble",""],["PreambleDesc",""],["PreambleInput",""],["PreambleInputDesc",""],["PreambleInputEnum",""],["PreambleInputEnumChange",""],["PreambleInputEnumDesc",""]]};