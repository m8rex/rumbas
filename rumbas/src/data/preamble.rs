use crate::data::file_reference::FileString;
use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::to_numbas::{NumbasResult, ToNumbas};
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    Preamble,
    js: FileString,
    css: FileString
}

impl ToNumbas for Preamble {
    type NumbasType = numbas::exam::Preamble;
    fn to_numbas(&self, _locale: &String) -> NumbasResult<numbas::exam::Preamble> {
        let empty_fields = self.empty_fields();
        if empty_fields.is_empty() {
            Ok(numbas::exam::Preamble::new(
                self.js.clone().unwrap().get_content(),
                self.css.clone().unwrap().get_content(),
            ))
        } else {
            Err(empty_fields)
        }
    }
}
