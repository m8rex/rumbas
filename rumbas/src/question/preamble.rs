use crate::support::file_reference::FileString;
use crate::support::optional_overwrite::*;
use crate::support::template::{Value, ValueType};
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    pub struct Preamble {
        /// The JavaScript to add to the outputfiles
        js: FileString,
        /// The CSS to add to the outputfiles
        css: FileString
    }
}

impl ToNumbas<numbas::exam::Preamble> for Preamble {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Preamble {
        numbas::exam::Preamble {
            js: self.js.to_numbas(locale),
            css: self.css.to_numbas(locale),
        }
    }
}

impl ToRumbas<Preamble> for numbas::exam::Preamble {
    fn to_rumbas(&self) -> Preamble {
        Preamble {
            js: self.js.to_rumbas(),
            css: self.css.to_rumbas(),
        }
    }
}
