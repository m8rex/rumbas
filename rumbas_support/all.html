<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../rumbas_support/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../rumbas_support/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location">Crate rumbas_support</h2></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../rumbas_support/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><h1 class="fqn"><span class="in-band">List of all items</span></h1><h3 id="Structs">Structs</h3><ul class="structs docblock"><li><a href="input/struct.FileToLoad.html">input::FileToLoad</a></li><li><a href="input/struct.InputCheckInvalidYamlData.html">input::InputCheckInvalidYamlData</a></li><li><a href="input/struct.InputCheckMissingData.html">input::InputCheckMissingData</a></li><li><a href="input/struct.InputCheckPath.html">input::InputCheckPath</a></li><li><a href="input/struct.InputCheckResult.html">input::InputCheckResult</a></li><li><a href="input/struct.LoadedLocalizedFile.html">input::LoadedLocalizedFile</a></li><li><a href="input/struct.LoadedNormalFile.html">input::LoadedNormalFile</a></li><li><a href="rumbas_check/struct.RumbasCheckInvalidJMEStringData.html">rumbas_check::RumbasCheckInvalidJMEStringData</a></li><li><a href="rumbas_check/struct.RumbasCheckMissingData.html">rumbas_check::RumbasCheckMissingData</a></li><li><a href="rumbas_check/struct.RumbasCheckPath.html">rumbas_check::RumbasCheckPath</a></li><li><a href="rumbas_check/struct.RumbasCheckResult.html">rumbas_check::RumbasCheckResult</a></li><li><a href="value/struct.TemplateString.html">value::TemplateString</a></li><li><a href="value/struct.TemplateStringDesc.html">value::TemplateStringDesc</a></li><li><a href="value/struct.Value.html">value::Value</a></li><li><a href="value/struct.ValueChange.html">value::ValueChange</a></li><li><a href="value/struct.ValueDesc.html">value::ValueDesc</a></li></ul><h3 id="Enums">Enums</h3><ul class="enums docblock"><li><a href="input/enum.LoadedFile.html">input::LoadedFile</a></li><li><a href="value/enum.TemplateStringChange.html">value::TemplateStringChange</a></li><li><a href="value/enum.ValueType.html">value::ValueType</a></li><li><a href="value/enum.ValueTypeChange.html">value::ValueTypeChange</a></li><li><a href="value/enum.ValueTypeDesc.html">value::ValueTypeDesc</a></li></ul><h3 id="Traits">Traits</h3><ul class="traits docblock"><li><a href="example/trait.Examples.html">example::Examples</a></li><li><a href="input/trait.Input.html">input::Input</a></li><li><a href="input/trait.InputInverse.html">input::InputInverse</a></li><li><a href="overwrite/trait.Overwrite.html">overwrite::Overwrite</a></li><li><a href="rumbas_check/trait.RumbasCheck.html">rumbas_check::RumbasCheck</a></li></ul><h3 id="Constants">Constants</h3><ul class="constants docblock"><li><a href="value/constant.TEMPLATE_PREFIX.html">value::TEMPLATE_PREFIX</a></li></ul></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="rumbas_support" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (f03ce3096 2022-08-08)" ></div></body></html>