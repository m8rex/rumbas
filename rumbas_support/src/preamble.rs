pub use crate::input::*;
pub use crate::overwrite::*;
pub use crate::rumbas_check::*;
pub use crate::value::*;
