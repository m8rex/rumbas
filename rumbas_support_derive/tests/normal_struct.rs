#[macro_use]
extern crate rumbas_support_derive;

use comparable::Comparable;
use rumbas_support::preamble::*;
use serde::Deserialize;
use serde::Serialize;

#[derive(Input, RumbasCheck, Examples)]
#[input(name = "TestInput")]
#[derive(Clone, Debug, Deserialize, Serialize, Comparable, PartialEq)]
pub struct Test {
    field1: bool,
    field2: f64,
}

type TestInputs = Vec<Test>;

#[derive(Input, RumbasCheck, Examples)]
#[input(name = "Test2Input")]
#[input(test)]
#[derive(Clone, Debug, Deserialize, Serialize, Comparable, PartialEq)]
pub struct Test2 {
    field1: TestInputs,
    field2: f64,
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "TestOverwriteInput")]
#[derive(Clone, Deserialize)]
pub struct TestOverwrite {
    field1: bool,
    field2: f64,
}

#[derive(Input, RumbasCheck, Examples)]
#[input(name = "TestFromAndIntoInput")]
#[input(from = "String")]
#[input(into = "String")]
#[derive(Clone, Debug, Deserialize, Serialize, Comparable, PartialEq, Eq)]
pub struct TestFromAndInto {
    // TODO: add real test
    field1: String,
    field2: String,
}

impl std::convert::From<String> for TestFromAndIntoInput {
    fn from(s: String) -> TestFromAndIntoInput {
        Self {
            field1: Value::Normal(s.clone()),
            field2: Value::Normal(s),
        }
    }
}

impl std::convert::From<TestFromAndIntoInput> for String {
    fn from(s: TestFromAndIntoInput) -> String {
        s.field1.unwrap()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn create_test2() {
        let _test2 = Test2 {
            field1: vec![Test {
                field1: true,
                field2: 64.8,
            }],
            field2: 65.0,
        };

        let _test2 = Test2Input {
            field1: Value::Normal(vec![ValueType::Normal(TestInput {
                field1: Value::Normal(true),
                field2: Value::Normal(64.8),
            })]),
            field2: Value::Normal(65.0),
        };
    }

    #[test]
    fn parse_yaml_test_no_fields_ok_for_normal() {
        let fail: Result<Test2Input, _> = serde_yaml::from_str(
            r"---
other_field1: true
",
        );
        assert!(fail.is_ok());
    }

    #[test]
    fn parse_yaml_test_no_fields() {
        let fail: Result<Test2InputEnum, _> = serde_yaml::from_str(
            r"---
other_field1: true
",
        );
        assert!(fail.is_err());
    }

    #[test]
    fn parse_yaml_test() {
        let ok: Result<Test2InputEnum, _> = serde_yaml::from_str(
            r"---
field1: true
",
        );

        assert!(ok.is_ok());
    }
    #[test]
    fn examples() {
        TestInput::examples();
        TestInputEnum::examples();
        Test2Input::examples();
        Test2InputEnum::examples();
    }
}
