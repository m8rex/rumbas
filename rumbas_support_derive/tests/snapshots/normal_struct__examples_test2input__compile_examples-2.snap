---
source: rumbas_support_derive/tests/normal_struct.rs
expression: "&example"

---
field1: "template:template_key"
field2: "template:template_key"

